extensions/amp-story/0.1/amp-story-info-dialog.js

```diff
@@ -0,0 +1,252 @@
+    const html = htmlFor(this.parentEl_);
+    this.element_ = html`
+        <div class="i-amphtml-story-info-dialog i-amphtml-story-system-reset">
+          <div class="i-amphtml-story-info-dialog-container">
+            <h1 class="i-amphtml-story-info-heading"></h1>
+            <a class="i-amphtml-story-info-link"></a>
+            <a class="i-amphtml-story-info-moreinfo"></a>
+          </div>
+        </div>`;
+
+    createShadowRootWithStyle(root, this.element_, CSS);
+    this.initializeListeners_();
+
+    this.innerContainerEl_ = this.element_
+        .querySelector('.i-amphtml-story-info-dialog-container');
+
+    const appendPromise = this.resources_.mutateElement(this.parentEl_, () => {
+      this.parentEl_.appendChild(root);
+    });
+
+    const pageUrl = Services.documentInfoForDoc(
+        getAmpdoc(this.parentEl_)).canonicalUrl;
+
+    return Promise.all([
+      appendPromise,
+      this.setHeading_(),
+      this.setPageLink_(pageUrl),
+      this.requestMoreInfoLink_()
+          .then(moreInfoUrl => this.setMoreInfoLinkUrl_(moreInfoUrl)),
+    ]);
+  }
+
+  /**
+   * Whether the element has been built.
+   * @return {boolean}
+   */
+  isBuilt() {
+    return this.isBuilt_;
+  }
+
+  /**
+   * @private
+   */
+  initializeListeners_() {
+    this.storeService_.subscribe(StateProperty.INFO_DIALOG_STATE, isOpen => {
+      this.onInfoDialogStateUpdated_(isOpen);
+    });
+
+    this.element_.addEventListener(
+        'click', event => this.onInfoDialogClick_(event));
+  }
+
+  /**
+   * Reacts to menu state updates and decides whether to show either the native
+   * system sharing, or the fallback UI.
+   * @param {boolean} isOpen
+   * @private
+   */
+  onInfoDialogStateUpdated_(isOpen) {
+    this.resources_.mutateElement(dev().assertElement(this.element_), () => {
+      this.element_.classList.toggle(DIALOG_VISIBLE_CLASS, isOpen);
+    });
+  }
+
+  /**
+   * Handles click events and maybe closes the dialog.
+   * @param  {!Event} event
+   */
+  onInfoDialogClick_(event) {
+    const el = dev().assertElement(event.target);
+    // Closes the dialog if click happened outside of the dialog main container.
+    if (!closest(el, el => el === this.innerContainerEl_, this.element_)) {
+      this.close_();
+    }
+  }
+
+  /**
+   * Closes the info dialog.
+   * @private
+   */
+  close_() {
+    this.storeService_.dispatch(Action.TOGGLE_INFO_DIALOG, false);
+  }
+
+  /**
+   * @return {!Promise<?string>} The URL to visit to receive more info on this
+   *     page.
+   * @private
+   */
+  requestMoreInfoLink_() {
+    const messagingPromise = this.viewer_.whenMessagingReady();
+
+    if (!messagingPromise) {
+      // There is no viewer to supply the more info URL.
+      return Promise.resolve();
+    }
+
+    return messagingPromise
+        .then(() => {
+          return this.viewer_.sendMessageAwaitResponse('moreInfoLinkUrl',
+              /* data */ undefined);
```

Comment:
Do we need the runtime to ask for this URL or can it be provided by the viewer at load time via hash param? See `viewer-impl.getParam()`.

Reply:
This was considered, but decided against to keep the URL length down, since we don't strictly need this populated when the document is initialized (since the user is likely to not click the button that renders this UI until after the messaging has completed).